packages/dev/core/test/unit/DeviceInput/babylon.deviceInput.test.ts

```diff
@@ -0,0 +1,226 @@
+        // Do nothing
+    }
+}
+
+class MockInternalDeviceSourceManager {
+    public _refCount = 0;
+
+    constructor() {}
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public readonly registerManager = (manager: IObservableManager): void => {
+        // Do nothing
+    };
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public readonly unregisterManager = (manager: IObservableManager): void => {
+        // Do nothing
+    };
+
+    public dispose(): void {}
+}
+
+describe("DeviceSource", () => {
+    it("should exist", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        expect(mouseSource).not.toBeNull();
+        expect(mouseSource.deviceType).toBe(DeviceType.Mouse);
+        expect(mouseSource.deviceSlot).toBe(0);
+        expect(mouseSource.onInputChangedObservable).not.toBeUndefined();
+    });
+    it("can poll with getInput", () => {
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, null);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        const leftClick = mouseSource.getInput(PointerInput.LeftClick);
+        expect(leftClick).toBe(1);
+    });
+    it("can use onInputChangedObservable", () => {
+        expect.assertions(2);
+        let resultEvent: Nullable<IWheelEvent | IPointerEvent> = null;
+        let onInputChanged = null;
+        // Create DeviceSource and add to its observable
+        const deviceInputSystem = new MockDeviceInputSystem(null, null, onInputChanged);
+        const mouseSource = new DeviceSource(deviceInputSystem, DeviceType.Mouse, 0);
+        onInputChanged = (deviceType: DeviceType, deviceSlot: number, eventData: IUIEvent) => {
+            mouseSource.onInputChangedObservable.notifyObservers(eventData as IPointerEvent);
+        };
+        mouseSource.onInputChangedObservable.add((eventData) => {
+            resultEvent = eventData;
+        });
+
+        expect(mouseSource.onInputChangedObservable.hasObservers()).toBe(true);
+        const mouseEvent: IUIEvent = { pointerId: 0, type: "pointerdown", pointerType: "mouse", clientX: 1, clientY: 1 } as IPointerEvent;
+        onInputChanged(DeviceType.Mouse, 0, mouseEvent);
+
+        expect(resultEvent).toEqual(mouseEvent);
+    });
+});
+
+describe("DeviceSourceManager", () => {
+    let engine: Nullable<NullEngine> = null;
+
+    beforeEach(function () {
+        engine = new NullEngine();
+        engine._deviceSourceManager = new MockInternalDeviceSourceManager() as InternalDeviceSourceManager;
+    });
+
+    afterEach(function () {
+        engine?.dispose();
+    });
+
+    it("should exist", () => {
+        const deviceSourceManager = new DeviceSourceManager(engine!);
+        expect(deviceSourceManager).not.toBe(null);
+        expect(engine?._deviceSourceManager?._refCount).toBe(1);

```

Comment:
I would recommend using `engine!` in most places. It is guaranteed to exist right?

Reply:
I'll remove all instances of engine?